File: App/ios/Plugin/CoronaSoomlaStore.m
//
//  SoomlaStore.m
//  Soomla for Corona
//
//  Created by Bruno Barbosa Pinheiro on 3/7/14.

#import "CoronaSoomlaStore.h"
#import "StoreConfig.h"
#import "SoomlaStore.h"
#import "VirtualItem.h"
#import "VirtualCurrency.h"
#import "VirtualCurrencyPack.h"
#import "VirtualCategory.h"
#import "VirtualGood.h"
#import "NonConsumableItem.h"
#import "StoreInfo.h"
#import "Soomla.h"

#define kStore_Version                  @"version"
#define kStore_VirtualGoods             @"virtualGoods"
#define kStore_VirtualCurrencies        @"virtualCurrencies"
#define kStore_VirtualCurrencyPacks     @"virtualCurrencyPacks"
#define kStore_VirtualCategories        @"virtualCategories"
#define kStore_NonConsumableItems       @"nonConsumableItems"
#define kStore_CustomSecret             @"CUSTOM_SECRET"

@interface CoronaSoomlaStore ()

@property (nonatomic,strong) NSMutableDictionary * vItems;
@property (nonatomic,strong) NSMutableDictionary * vCategories;

@end

@implementation CoronaSoomlaStore

#pragma mark - Initialization

+ (SoomlaStore *) sharedInstance {
    static SoomlaStore * instance = nil;
    if(instance == nil) instance = [[SoomlaStore alloc] init];
    return instance;
}

- (id) init {
    self = [super init];
    if(self == nil) return nil;
    self.vItems = [[NSMutableDictionary alloc] init];
    self.vCategories = [[NSMutableDictionary alloc] init];
    self.avaiableCategories = [[NSArray alloc] init];
    self.avaiableCurrencies = [[NSArray alloc] init];
    self.avaiableCurrencyPacks = [[NSArray alloc] init];
    self.avaiableNonConsumableItems = [[NSArray alloc] init];
    self.avaiableVirtualGoods = [[NSArray alloc] init];
    return self;
}

- (void) initializeWithData:(NSDictionary *) luaData {
    NSNumber * version = [luaData objectForKey:kStore_Version];
    self.version = [version intValue];
    
    NSDictionary * categories = [luaData objectForKey:kStore_VirtualCategories];
    self.avaiableCategories = [categories allValues];
    
    NSDictionary * virtualCurrencies = [luaData objectForKey:kStore_VirtualCurrencies];
    self.avaiableCurrencies = [virtualCurrencies allValues];
    
    NSDictionary * virtualCurrencyPacks = [luaData objectForKey:kStore_VirtualCurrencyPacks];
    self.avaiableCurrencyPacks = [virtualCurrencyPacks allValues];
    
    NSDictionary * nonConsumableItems = [luaData objectForKey:kStore_NonConsumableItems];
    self.avaiableNonConsumableItems = [nonConsumableItems allValues];
    
    NSDictionary * virtualGoods = [luaData objectForKey:kStore_VirtualGoods];
    self.avaiableVirtualGoods = [virtualGoods allValues];
    
    NSString * customSecret = [luaData objectForKey:kStore_CustomSecret];
    [Soomla initializeWithSecret:customSecret];
    [[SoomlaStore getInstance] initializeWithStoreAssets:self];
}

//TODO: Return the correct array for each method

#pragma mark - Version

- (int) getVersion {
    return self.version;
}

- (void) addVirtualItem:(VirtualItem *) virtualItem {
    [self.vItems setObject:virtualItem forKey:virtualItem.itemId];
}

#pragma mark - Currency

- (NSArray *) virtualCurrencies {
    NSMutableArray * currencies = [[NSMutableArray alloc] init];
    for(NSString * currencyId in self.avaiableCurrencies) {
        VirtualCurrency * currency = (VirtualCurrency *)[self.vItems objectForKey:currencyId];
        if(currency == nil) NSLog(@"SOOMLA: %@ is not a Virtual Currency!! It will not be used!",currencyId);
        else [currencies addObject:currency];
    }
    return [NSArray arrayWithArray:currencies];
}

- (NSArray *) virtualCurrencyPacks {
    NSMutableArray * currencyPacks = [[NSMutableArray alloc] init];
    for(NSString * currencyPackId in self.avaiableCurrencyPacks) {
        VirtualCurrencyPack * currencyPack = (VirtualCurrencyPack *)[self.vItems objectForKey:currencyPackId];
        if(currencyPack == nil) NSLog(@"SOOMLA: %@ is not a Virtual Currency Pack!! It will not be used!",currencyPackId);
        else [currencyPacks addObject:currencyPack];
    }
    return [NSArray arrayWithArray:currencyPacks];
}

#pragma mark - Categories

- (void) addVirtualCategory:(VirtualCategory *) category {
    [self.vCategories setValue:category forKey:category.name];
}

- (VirtualCategory *) categoryWithName:(NSString *) name {
    return [self.vCategories objectForKey:name];
}

- (NSArray *) virtualCategories {
    NSMutableArray * categories = [[NSMutableArray alloc] init];
    for(NSString * categoryName in self.avaiableCategories) {
        VirtualCategory * category = (VirtualCategory *)[self.vCategories objectForKey:categoryName];
        if(category == nil) NSLog(@"SOOMLA: Category %@ doesn't exist!",categoryName);
        else [categories addObject:category];
    }
    return [NSArray arrayWithArray:categories];
}


#pragma mark - Virtual Goods

- (NSArray *) virtualGoods {
    NSMutableArray * virtualGoods = [[NSMutableArray alloc] init];
    for(NSString * virtualGoodId in self.avaiableVirtualGoods) {
        VirtualGood * virtualGood = (VirtualGood *)[self.vItems objectForKey:virtualGoodId];
        if(virtualGood == nil) NSLog(@"SOOMLA: %@ is not a Virtual Good!! It will not be used!",virtualGoodId);
        else [virtualGoods addObject:virtualGood];
    }
    return [NSArray arrayWithArray:virtualGoods];
}

#pragma mark - Non Consumable Items

- (NSArray *) nonConsumableItems {
    NSMutableArray * nonConsumableItems = [[NSMutableArray alloc] init];
    for(NSString * nonConsumableItemId in self.avaiableNonConsumableItems) {
        NonConsumableItem * nonConsumableItem = (NonConsumableItem *)[self.vItems objectForKey:nonConsumableItemId];
        if(nonConsumableItem == nil) NSLog(@"SOOMLA: %@ is not a Non Consumable Item!! It will not be used!",nonConsumableItemId);
        else [nonConsumableItems addObject:nonConsumableItem];
    }
    return [NSArray arrayWithArray:nonConsumableItems];
}


@end

File: App/android/src/com/soomla/corona/store/CoronaSoomlaStore.java
package com.soomla.corona.store;

import com.soomla.BusProvider;
import com.soomla.Soomla;
import com.soomla.store.billing.google.*;
import com.soomla.store.IStoreAssets;
import com.soomla.store.SoomlaStore;
import com.soomla.store.domain.*;
import com.soomla.store.domain.virtualCurrencies.*;
import com.soomla.store.domain.virtualGoods.*;

import java.lang.Integer;
import java.lang.Override;
import java.lang.String;
import java.lang.Exception;
import java.util.Map;
import java.util.HashMap;
import java.util.ArrayList;

public class CoronaSoomlaStore implements IStoreAssets {

    /// Singleton
    private CoronaSoomlaStore() {}

    public static CoronaSoomlaStore getInstance() {
        if(_instance == null) _instance = new CoronaSoomlaStore();
        return _instance;
    }

    private static CoronaSoomlaStore _instance;

    /// Initialization
    public static final String VERSION          = "version";
    public static final String CATEGORIES       = "virtualCategories";
    public static final String CURRENCIES       = "virtualCurrencies";
    public static final String CURRENCYPACKS    = "virtualCurrencyPacks";
    public static final String VIRTUALGOODS     = "virtualGoods";
    public static final String NONCONSUMABLE    = "nonConsumableItems";
    public static final String CUSTOMSECRET     = "CUSTOM_SECRET";
    public static final String GOOGLEPLAYKEY    = "GOOGLE_PLAY_KEY";


    public void initialize(Map<String,Object> map) throws Exception {
        try {
            //Making our EventListener start doing its job!
            BusProvider.getInstance().register(EventListener.getInstance());

            this.version = ((Double)map.get(CoronaSoomlaStore.VERSION)).intValue();

            Map<String,String> categories = (Map<String,String>)map.get(CoronaSoomlaStore.CATEGORIES);
            this.availableCategories = new ArrayList<String>(categories.values());

            Map<String,String> currencies = (Map<String,String>)map.get(CoronaSoomlaStore.CURRENCIES);
            this.availableCurrencies = new ArrayList<String>(currencies.values());

            Map<String,String> currencyPacks = (Map<String,String>)map.get(CoronaSoomlaStore.CURRENCYPACKS);
            this.availableCurrencyPacks = new ArrayList<String>(currencyPacks.values());

            Map<String,String> virtualGoods = (Map<String,String>)map.get(CoronaSoomlaStore.VIRTUALGOODS);
            this.availableVirtualGoods = new ArrayList<String>(virtualGoods.values());

            Map<String,String> nonConsumableItems = (Map<String,String>)map.get(CoronaSoomlaStore.NONCONSUMABLE);
            this.availableNonConsumableItems = new ArrayList<String>(nonConsumableItems.values());

            this.customSecret = (String)map.get(CoronaSoomlaStore.CUSTOMSECRET);
            //this.googlePlayKey = (String)map.get(CoronaSoomlaStore.GOOGLEPLAYKEY);

            System.out.println("SOOMLA: Initializing the custom secret...");
            Soomla.initialize(this.customSecret);
            System.out.println("SOOMLA: Initializing the IStoreAssets...");
            SoomlaStore.getInstance().initialize(this);
            System.out.println("SOOMLA: Initializing the Google Play IAB Service");
            //GooglePlayIabService.getInstance().setPublicKey(this.googlePlayKey);
        } catch(Exception e) { throw e; }
    }

    /// Version
    @Override public int getVersion() {
        return this.version;
    }

    private int version = 0;

    /// Security
    private String customSecret = "UNKNOWN";
    private String googlePlayKey = "UNKNOWN";

    /// Virtual Items
    public void addVirtualItem(VirtualItem virtualItem) {
        this.vItems.put(virtualItem.getItemId(),virtualItem);
    }

    private HashMap<String,VirtualItem> vItems = new HashMap<String,VirtualItem>();


    /// Category
    public void addVirtualCategory(VirtualCategory category) {
        this.vCategories.put(category.getName(),category);
    }

    public VirtualCategory getCategory(String name) {
        return this.vCategories.get(name);
    }

    private HashMap<String,VirtualCategory> vCategories = new HashMap<String,VirtualCategory>();

    /// Availables
    @Override public VirtualCurrency[] getCurrencies() {
        ArrayList<VirtualCurrency> currencies = new ArrayList<VirtualCurrency>();
        for(String currencyId : this.availableCurrencies) {
            VirtualCurrency currency = (VirtualCurrency)this.vItems.get(currencyId);
            currencies.add(currency);
        }
        VirtualCurrency[] currencyArray = new VirtualCurrency[currencies.size()];
        return currencies.toArray(currencyArray);
    }

    @Override public VirtualCurrencyPack[] getCurrencyPacks() {
        ArrayList<VirtualCurrencyPack> currencyPacks = new ArrayList<VirtualCurrencyPack>();
        for(String currencyPackId : this.availableCurrencyPacks) {
            VirtualCurrencyPack currencyPack = (VirtualCurrencyPack)this.vItems.get(currencyPackId);
            currencyPacks.add(currencyPack);
        }
        VirtualCurrencyPack[] currencyPackArray = new VirtualCurrencyPack[currencyPacks.size()];
        return currencyPacks.toArray(currencyPackArray);
    }

    @Override public VirtualCategory[] getCategories() {
        ArrayList<VirtualCategory> categories = new ArrayList<VirtualCategory>();
        for(String categoryName : this.availableCurrencies) {
            VirtualCategory category = this.vCategories.get(categoryName);
            categories.add(category);
        }
        VirtualCategory[] categoryArray = new VirtualCategory[categories.size()];
        return categories.toArray(categoryArray);
    }

    @Override public VirtualGood[] getGoods() {
        ArrayList<VirtualGood> virtualGoods = new ArrayList<VirtualGood>();
        for(String virtualGoodId : this.availableVirtualGoods) {
            VirtualGood virtualGood = (VirtualGood)this.vItems.get(virtualGoodId);
            virtualGoods.add(virtualGood);
        }
        VirtualGood[] virtualGoodArray = new VirtualGood[virtualGoods.size()];
        return virtualGoods.toArray(virtualGoodArray);
    }

    @Override public NonConsumableItem[] getNonConsumableItems() {
        ArrayList<NonConsumableItem> nonConsumableItems = new ArrayList<NonConsumableItem>();
        for(String nonConsumableItemId : this.availableNonConsumableItems) {
            NonConsumableItem nonConsumableItem = (NonConsumableItem)this.vItems.get(nonConsumableItemId);
            nonConsumableItems.add(nonConsumableItem);
        }
        NonConsumableItem[] nonConsumableItemArray = new NonConsumableItem[nonConsumableItems.size()];
        return nonConsumableItems.toArray(nonConsumableItemArray);
    }


    private ArrayList<String> availableCategories = new ArrayList<String>();
    private ArrayList<String> availableCurrencies = new ArrayList<String>();
    private ArrayList<String> availableCurrencyPacks = new ArrayList<String>();
    private ArrayList<String> availableNonConsumableItems = new ArrayList<String>();
    private ArrayList<String> availableVirtualGoods = new ArrayList<String>();
}